File: sample/src/main/java/io/noties/adapt/sample/NextActivity.java
package io.noties.adapt.sample;

import android.app.Activity;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Calendar;
import java.util.Collections;
import java.util.Date;

import io.noties.adapt.next.Adapt;
import io.noties.adapt.next.Item;
import io.noties.adapt.next.ItemList;
import io.noties.adapt.next.listview.AdaptListView;
import io.noties.adapt.next.listview.ListViewItem;
import io.noties.adapt.next.recyclerview.AdaptRecyclerView;
import io.noties.adapt.next.recyclerview.AsyncDiffUtilDataSetChangedHandler;
import io.noties.adapt.next.recyclerview.DiffUtilDataSetChangedHandler;

public class NextActivity extends Activity {
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

//        final Adapt adapt = listView();
        final Adapt adapt = recyclerView();

        findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {

            int count = 0;

            @Override
            public void onClick(View v) {
                if (count == 0) {
                    final ItemList list = new ItemList();
                    for (int i = 0; i < 10; i++) {
                        list.add(new StringItem("String #" + (i + 1)));
                    }
                    adapt.setItems(list);
                } else if (count == 1) {
                    final ItemList list = new ItemList(adapt.items());
                    for (int i = 0; i < 20; i++) {
                        list.add(new NumberItem(i + 1));
                    }
                    Collections.shuffle(list);
                    adapt.setItems(list);
                } else if (count == 2) {
                    final Calendar calendar = Calendar.getInstance();
                    final ItemList list = new ItemList(adapt.items());
                    for (int i = 0; i < 30; i++) {
                        calendar.add(Calendar.DAY_OF_MONTH, i);
                        list.add(new DateItem(calendar.getTime(), new DateItem.OnClick() {
                            @Override
                            public void onClick(@NonNull Date date) {
                                Toast.makeText(NextActivity.this, date.toString(), Toast.LENGTH_LONG).show();
                            }
                        }));
                    }
                    Collections.shuffle(list);
                    adapt.setItems(list);
                } else {
                    final ItemList list = new ItemList(adapt.items());
                    Collections.shuffle(list);
                    adapt.setItems(list);
                }

                count += 1;
            }
        });
    }

    @NonNull
    private Adapt listView() {
        setContentView(R.layout.activity_list_view);

        final ListView listView = findViewById(R.id.list_view);

        return AdaptListView.attach(listView, new AdaptListView.Configurator() {
            @Override
            public void configure(@NonNull AdaptListView.Configuration configuration) {
                configuration.areAllItemsEnabled(false);
            }
        });
    }

    @NonNull
    private Adapt recyclerView() {
        setContentView(R.layout.activity_recycler_view);

        final RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setHasFixedSize(true);

        return AdaptRecyclerView.attach(recyclerView, new AdaptRecyclerView.Configurator() {
            @Override
            public void configure(@NonNull AdaptRecyclerView.Configuration configuration) {
                configuration.dataSetChangeHandler(AsyncDiffUtilDataSetChangedHandler.create(
                        DiffUtilDataSetChangedHandler.create(true)
                ));
            }
        });
    }

    private static class StringItem extends Item<StringItem.Holder> implements ListViewItem {

        private final String string;

        protected StringItem(@NonNull String string) {
            super(string.hashCode());
            this.string = string;
        }

        @NonNull
        @Override
        public Holder createHolder(@NonNull LayoutInflater inflater, @NonNull ViewGroup parent) {
            return new Holder(inflater.inflate(android.R.layout.simple_list_item_1, parent, false));
        }

        @Override
        public void render(@NonNull Holder holder) {
            holder.textView.setText(string);
        }

        @Override
        public boolean listViewIsEnabled() {
            return false;
        }

        static class Holder extends Item.Holder {

            final TextView textView;

            Holder(@NonNull View itemView) {
                super(itemView);
                textView = requireView(android.R.id.text1);
            }
        }
    }

    private static class NumberItem extends Item<NumberItem.Holder> {

        private final int number;

        protected NumberItem(int number) {
            super(number);
            this.number = number;
        }

        @NonNull
        @Override
        public Holder createHolder(@NonNull LayoutInflater inflater, @NonNull ViewGroup parent) {
            return new Holder(inflater.inflate(android.R.layout.simple_list_item_2, parent, false));
        }

        @Override
        public void render(@NonNull Holder holder) {
            holder.text1.setText(String.valueOf(number));
            holder.text2.setText("Number");
        }

        static class Holder extends Item.Holder {

            final TextView text1;
            final TextView text2;

            Holder(@NonNull View itemView) {
                super(itemView);
                text1 = requireView(android.R.id.text1);
                text2 = requireView(android.R.id.text2);
            }
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            NumberItem that = (NumberItem) o;

            return number == that.number;
        }

        @Override
        public int hashCode() {
            return number;
        }
    }

    private static class DateItem extends Item<DateItem.Holder> {

        interface OnClick {
            void onClick(@NonNull Date date);
        }

        private final Date date;
        private final OnClick onClick;

        private DateItem(@NonNull Date date, @NonNull OnClick onClick) {
            super(date.getTime());
            this.date = date;
            this.onClick = onClick;
        }

        @NonNull
        @Override
        public Holder createHolder(@NonNull LayoutInflater inflater, @NonNull ViewGroup parent) {
            final Holder holder = new Holder(inflater.inflate(android.R.layout.simple_list_item_1, parent, false));
            holder.itemView().setBackgroundColor(Color.YELLOW);
            return holder;
        }

        @Override
        public void render(@NonNull Holder holder) {
            holder.textView.setText(date.toString());
            holder.itemView().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onClick.onClick(date);
                }
            });
        }

        static class Holder extends Item.Holder {
            final TextView textView;

            Holder(@NonNull View itemView) {
                super(itemView);
                this.textView = requireView(android.R.id.text1);
            }
        }
    }
}

File: app/src/main/java/ru/noties/adapt/sample/next/NextActivity.kt
package ru.noties.adapt.sample.next

import android.app.Activity
import android.graphics.Rect
import android.os.Bundle
import android.os.Handler
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.transition.TransitionManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import ru.noties.adapt.next.*
import ru.noties.adapt.sample.R
import ru.noties.adapt.sample.core.ColorRandom
import ru.noties.adapt.sample.core.ShapeRandom
import ru.noties.debug.AndroidLogDebugOutput
import ru.noties.debug.Debug

class NextActivity : Activity() {

    private var index = 0
    private var update = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_next)

        Debug.init(AndroidLogDebugOutput(true))

        val recycler = findViewById<RecyclerView>(R.id.recycler_view)
        if (true) {

            val adapt = Adapt.create(AsyncDiffUtilDataSetChanged.create())

            recycler.layoutManager = LinearLayoutManager(this)
            recycler.setHasFixedSize(true)
            recycler.adapter = adapt

            val shapeRandom = ShapeRandom.create()
            val colorRandom = ColorRandom.create()

            val items = listOf<Item<*>>(
                    IntItem(10),
                    CharItem('z'),
                    MyItem(shapeRandom.next(), colorRandom.next(), "0", "Subtitle here")
            )
            adapt.setItems(items)

            val handler = Handler()
            handler.postDelayed({
                (0..10)
                        .map { MyItem(shapeRandom.next(), colorRandom.next(), "$it", "Just some text on top") }
                        .toList()
                        .apply {
                            adapt.setItems(this)
                            adapt.setItems(this)
                            adapt.setItems(this)
                        }

            }, 2000L)

            return
        }
        val parent = findViewById<ViewGroup>(/*R.id.group*/0)

        if (true) {

            val shapeRandom = ShapeRandom.create()
            val colorRandom = ColorRandom.create()
            val items = mutableListOf<MyItem>()
            val adaptViews = mutableListOf<AdaptView<MyItem>>()
            val handler = Handler()
            handler.postDelayed(object : Runnable {
                override fun run() {

                    if (!update) {

                        TransitionManager.beginDelayedTransition(parent)

                        val item = MyItem(
                                shapeRandom.next(),
                                colorRandom.next(),
                                "Item #$index",
                                "Super duper huge big subtitle for item #$index")
                        val view = AdaptView.append(parent, item)
                        adaptViews.add(view)
                        items.add(item)
                    } else {

//                        TransitionManager.beginDelayedTransition(parent)

                        val item = items[index].run {
                            copy(
                                    shapeType = shapeRandom.next(),
                                    color = colorRandom.next(),
                                    subtitle = "${this.subtitle} + updated!")
                        }
                        val view = adaptViews[index]

//                        view.bind(MyItem(shapeRandom.next(), colorRandom.next(), item.title, "${item.subtitle} + updated!"))
                        view.bind(item)
//                        AdaptView.create(
//                                parent.getChildAt(index),
//                                MyItem(shapeRandom.next(), colorRandom.next(), item.title, "${item.subtitle} + updated!")) {
//                            MyItem.Holder(it)
//                        }
                    }

                    if (index == 4 && update) {

                        TransitionManager.beginDelayedTransition(parent)

                        val group = AdaptViewGroup.create(parent)
                        group.setItems(
                                items.map {
                                    it.copy(
                                            shapeRandom.next(),
                                            colorRandom.next(),
                                            it.title,
                                            "${it.subtitle} + uououououou!"
                                    )
                                }
                                        .reversed()
                        )
                        return
                    }

                    index += 1

                    if (index == 5) {
                        index = 0
                        update = true
                    }

                    handler.postDelayed(this, 1000L)
                }
            }, 1000L)

            return
        }

        val steps = arrayListOf<List<Item<*>>>(
                listOf(IntItem(10)),
                listOf(IntItem(10), IntItem(11)),
                listOf(IntItem(11), IntItem(10)),
                listOf(CharItem('a'), CharItem('b'), IntItem(11), IntItem(10)),
                listOf(IntItem(10), CharItem('a'), CharItem('b'), IntItem(11)),
                listOf(CharItem('a'), CharItem('b'), IntItem(11))
        )

        if (true) {

            val adaptViewGroup = AdaptViewGroup.create(parent)
            val handler = Handler()
            val runnable = object : Runnable {
                override fun run() {

                    TransitionManager.beginDelayedTransition(parent)

                    val start = System.currentTimeMillis()
//                    NextDiff.diff(parent, previous, steps[index])
                    adaptViewGroup.setItems(steps[index])
                    val end = System.currentTimeMillis()
                    Debug.i("took: %d ms", end - start);

                    index += 1
                    if (index < steps.size) {
//                        previous.clear()
//                        previous.addAll(steps[index - 1])
                        handler.postDelayed(this, 1000L)
                    }
                }

            }
            handler.postDelayed(runnable, 1000L)
            return
        }

        var previous = mutableListOf<Item<*>>()

        val handler = Handler()
        val runnable = object : Runnable {
            override fun run() {

                TransitionManager.beginDelayedTransition(parent)

                val start = System.currentTimeMillis()
                NextDiff.diff(parent, previous, steps[index])
                val end = System.currentTimeMillis()
                Debug.i("took: %d ms", end - start);

                index += 1
                if (index < steps.size) {
                    previous.clear()
                    previous.addAll(steps[index - 1])
                    handler.postDelayed(this, 1000L)
                }
            }
        }

        handler.postDelayed(runnable, 1000L)
    }

    class IntItem(private val value: Int) : Item<IntItem.Holder>(value.toLong()) {

        override fun createHolder(inflater: LayoutInflater, parent: ViewGroup): Holder {
            return Holder(inflater.inflate(R.layout.adapt_int, parent, false))
        }

        override fun render(holder: Holder) {
            holder.textView.text = value.toString()
        }

        override fun recyclerDecoration(): RecyclerView.ItemDecoration? {
            return object : RecyclerView.ItemDecoration() {
                override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
                    outRect.set(48, 48, 48, 48)
//                    outRect.set(0, 0, 0, 0)
//                    val holder = parent.findContainingViewHolder(view) ?: return
//                    if (recyclerViewType() == holder.itemViewType) {
//                        outRect.set(48, 48, 48, 48)
//                    }
                }
            }
        }

        class Holder(itemView: View) : Item.Holder(itemView) {
            val textView = requireView<TextView>(R.id.text)
        }
    }

    class CharItem(private val value: Char) : Item<CharItem.Holder>(value.toLong()) {

        override fun createHolder(inflater: LayoutInflater, parent: ViewGroup): Holder {
            return Holder(inflater.inflate(R.layout.adapt_char, parent, false))
        }

        override fun render(holder: Holder) {
            holder.textView.text = value.toString()
        }

        class Holder(itemView: View) : Item.Holder(itemView) {
            val textView = requireView<TextView>(R.id.text)
        }
    }
}